### PR TITLE
Switch to clingo pypi package

### PR DESCRIPTION
**Pull request by anubhav-cs**
Alters the clingo_wrapper,  removing the requirement to manually install clingo(gringo) binary.

### CODE REVIEW

**Comment by miquelramirez** [comment]
Hi @anubhav-cs,

thanks for this. I am merging into `aig-upf:devel`.

Miquel.

**Comment by haz** [comment]
What was the motivation behind this? I'm worried that we may face issues wrapping this up for in-browser use...

**Comment by miquelramirez** [comment]
What says on the tin

>  removing the requirement to manually install clingo(gringo) binary.

**Comment by haz** [comment]
Aye, but just out of convenience? It the same clingo binary that ships via pip? If so, why not just locate it differently and have a two-line change to nab the path right and include in the requirements? I think I'm missing some broader consequence here. Any new python package requires rebuilding that dependency (and following the chain of what _it_ requires), so potentially a major setback for browser embeds.

**Comment by miquelramirez** [comment]
> It the same clingo binary that ships via pip?

Unless some third party is impersonating the `Pypi` repository, I would assume so.

> If so, why not just locate it differently and have a two-line change to nab the path right and include in the requirements?

Because it is not the (de facto) standard way of handling packages in `Python`. All the PEPs and recommendations I  see only strongly suggest to use `Pypi` whenever available over alternative methods.

>  I think I'm missing some broader consequence here.

If I were you I would consider the possibility that there is no such consequence.

> Any new python package requires rebuilding that dependency (and following the chain of what it requires), so potentially a major setback for browser embeds.

IIRC the "tarski in the browser" use case 1) Windows-only and 2) depending heavily on Anaconda? Is that correct?

In any case, the `gringo` dependency should probably an extra dependency (e.g. you only get it if you run the command `pip install tarski[gringo]`). That's a _very simple_ change to request with _exactly 0_ drama attached to it.

**Comment by haz** [comment]
Nah, not the use-case and it (gringo) is actually something we use (for grounding). Concretely, being able to get it all going in the browser is what enabled [this](https://mulab.ai/publication/2021-icapsdemo-aspinall/) award-winning plugin, and opens the door to smart grounding client-side in the browser for anyone using the online editor.

**Comment by miquelramirez** [comment]
So little drama that here is the change 1d4ab534c613035d3e63f60f04e648d5f6e932a3

If there's a subset of `tarski` dependencies which should be made optional to support delicate environments @haz, please make a list and a RFC. Making a PR with the "optimal" dependency package is also helpful (but of course, I retain the right to reject or amend the PR as it suits).

@anubhav-cs could you please check that the `pip install tarski[gringo]` command is working as expected on your development environment (it should, see [docs here](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html)).

**Comment by haz** [comment]
Still doesn't help...the option is required in this case :P

**Comment by miquelramirez** [comment]
Then there you have your new default way of installing `tarski` if the `Pypi` solution doesn't suit. Nobody stops you or anybody from installing `tarski` without Pypi gringo, and then installing `gringo` by hand. Having it protected via the `extra_requires` method "shields" any award winning plugins from getting into trouble.

**Comment by haz** [comment]
The usage has fundamentally changed (e.g., [here](https://github.com/aig-upf/tarski/pull/123/files#diff-9d7875cf5ee765d0e41eaf0a8e137fc1ce9da88346662b8e3b3f5f02ab6ad8d6R23)), no? Are you suggesting that a non-gringo tarski with gringo binary added manually should "just work" after this merge?

**Comment by anubhav-cs** [comment]
> @anubhav-cs could you please check that the `pip install tarski[gringo]` command is working as expected on your development environment (it should, see [docs here](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html)).

@miquelramirez Yes it works.

**Comment by miquelramirez** [comment]
Well, that's an entirely different matter @haz.

@anubhav-cs you will need to add as a fallback the old `which`-based method. My suggestion is that you have a method/function called `get_gringo_command` that creates the arguments to pass to `cmd.execute`.  Just send another PR towards `devel`. Thank you.

**Comment by anubhav-cs** [comment]
@miquelramirez Easy peasy, will create a pull request soon.

**Comment by haz** [comment]
Merci both!

We could have potentially just frozen the version we include, but that's no fun...means all the latest/greatest stays out of the browser and editor.

**Comment by anubhav-cs** [comment]
I have created the PR https://github.com/aig-upf/tarski/pull/124 with the fallback method.

File: src/tarski/reachability/gringo.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from clingo.application import Application, clingo_main
+
+"""
+A wrapper to emulate the default clingo app behavior
+"""
+class WrapperClingo(Application):
+    def __init__(self, name):
+        self.app_name = name
+
+    def main(self, ctl, files):
```
Comment (gfrances) [inline src/tarski/reachability/gringo.py:12]: Innocent question: do we need all this code if we're just simply using the default behavior of clingo?

File: src/tarski/reachability/gringo.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
+import sys
+from clingo.application import Application, clingo_main
+
+"""
+A wrapper to emulate the default clingo app behavior
+"""
+class WrapperClingo(Application):
+    def __init__(self, name):
+        self.app_name = name
```
Comment (gfrances) [inline src/tarski/reachability/gringo.py:10]: Is this attribute being used anywhere?

Comment (anubhav-cs) [reply src/tarski/reachability/gringo.py:10]: That attribute should be `program_name`. That's a typo on my part. It is a member of the `Application` class.

File: src/tarski/reachability/gringo.py (hunk 1 above)
Comment (anubhav-cs) [reply src/tarski/reachability/gringo.py:12]: I had the same question. Going by the documentation this seems to be the recommended way to get the default clingo app behavior https://potassco.org/clingo/python-api/current/clingo/application.html 

Also, I checked it. The pip package doesn't include the gringo binary,  instead it has a platform specific library with a python wrapper. So, we cannot nab the path to the binary.

**Pull request merged**